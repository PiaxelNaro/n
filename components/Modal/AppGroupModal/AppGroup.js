// @flow
import React from 'react';
import styled from 'styled-components';
import FaCheckCircle from 'react-icons/lib/fa/check-circle-o';
import MdPlusCircle from 'react-icons/lib/md/control-point';
import Button from '../../Button';
import FlexGrid from '../../FlexGrid';
import Text from '../../Text';
import Spinner from '../../Spinner';
import { sprintf, type I18nType } from '../../../modules/i18n';

export type AppGroupObject = {
    id: number,
    link: string,
    name: string,
    apps: Array<{ id: number }>,
};

const StyledGroupWrapper = styled.li`
    padding: 5px 15px;
    height: 30px;
    line-height: 30px;
    color: #102235;
    font-size: 13px;
    :hover {
        background-color: whiteSmoke;
        .action-context {
           display: none;
        }
        .action-button {
           display: block;
        }
    }

    .action-button {
        display: none;
    }
`;

const Link = styled.a`
    color: #102235;
    text-decoration: none;
`;

const StyledGroupAppsCount = styled.span`
    margin: 0px 5px;
    color: #797979;
    white-space: nowrap;
`;

const AppGroup = ({
    group,
    appIds,
    addToGroup,
    isLoading = false,
    i18n,
}: {
    group: AppGroupObject,
    appIds: Array<number>,
    addToGroup: (number, Array<number>) => void,
    isLoading?: boolean,
    i18n: I18nType,
}) => {
    const { gettext, ngettext } = i18n;
    const selectedAppsCount = appIds.length;
    const addedAppsCount = group.apps.filter(app => appIds.includes(app.id)).length;
    let groupAction = null;
    const addAppsToGroups = () => addToGroup(group.id, appIds);

    /**
     * Three conditions of adding selected apps to group
     *  1. None added
     *  2. All Added
     *  3. Partial added
     */
    if (isLoading) {
        groupAction = <Spinner size={16} />;
    } else if (!addedAppsCount) {
        groupAction = (
            <div>
                <span className="action-context">
                    <MdPlusCircle size="16" color="#3F88D4" />
                </span>
                <span className="action-button">
                    <Button.Secondary onClick={addAppsToGroups}>
                        {gettext('Add To Group')}
                    </Button.Secondary>
                </span>
            </div>
        );
    } else if (selectedAppsCount === addedAppsCount) {
        groupAction = (
            <div>
                <FlexGrid inline className="action-context">
                    <StyledGroupAppsCount>
                        {gettext('All added')}
                    </StyledGroupAppsCount>
                    <FaCheckCircle size="16" color="#00BFAA" />
                </FlexGrid>
                <span className="action-button">
                    <Button.Secondary href={group.link}>
                        {gettext('Go To Group')}
                    </Button.Secondary>
                </span>
            </div>
        );
    } else {
        groupAction = (
            <div>
                <FlexGrid inline className="action-context">
                    <StyledGroupAppsCount>
                        {sprintf(gettext('%d already added'), addedAppsCount)}
                    </StyledGroupAppsCount>
                    <MdPlusCircle size="16" color="#3F88D4" />
                </FlexGrid>
                <span className="action-button">
                    <Button.Secondary onClick={addAppsToGroups}>
                        {gettext('Add Remaining')}
                    </Button.Secondary>
                </span>
            </div>
        );
    }

    return (
        <StyledGroupWrapper>
            <FlexGrid>
                <FlexGrid.Item>
                    <FlexGrid>
                        <Text ellipsis>
                            <Link href={group.link}>
                                {group.name}
                            </Link>
                        </Text>
                        <StyledGroupAppsCount>
                            {sprintf(
                                ngettext('(%d app)', '(%d apps)', group.apps.length),
                                group.apps.length
                            )}
                        </StyledGroupAppsCount>
                    </FlexGrid>
                </FlexGrid.Item>
                {groupAction}
            </FlexGrid>
        </StyledGroupWrapper>
    );
};

export default AppGroup;
