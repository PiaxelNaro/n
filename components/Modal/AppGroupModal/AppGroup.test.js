// @flow
import React from 'react';
import { shallow } from 'enzyme';
import AppGroup from './AppGroup';
import mockI18n from '../../../modules/i18n/mockI18n';

describe('<AppGroup>', () => {
    it('renders enabled group', () => {
        const fn = jest.fn();
        const groupList = shallow(
            <AppGroup
                appIds={[3, 4]}
                group={{
                    id: 1,
                    name: 'Group 1',
                    link: '/group-1',
                    apps: [{ id: 1 }, { id: 2 }],
                }}
                addToGroup={fn}
                i18n={mockI18n}
            />
        );
        expect(groupList).toMatchSnapshot();

        groupList.find('ButtonSecondary').simulate('click');
        expect(fn).toHaveBeenCalledWith(1, [3, 4]);
    });

    it('renders in loading mode', () => {
        expect(
            shallow(
                <AppGroup
                    isLoading
                    appIds={[3, 4]}
                    group={{
                        id: 1,
                        name: 'Group 1',
                        link: '/group-1',
                        apps: [{ id: 1 }, { id: 2 }],
                    }}
                    addToGroup={jest.fn()}
                    i18n={mockI18n}
                />
            )
        ).toMatchSnapshot();
    });

    it('renders in partially added mode', () => {
        expect(
            shallow(
                <AppGroup
                    appIds={[3, 4]}
                    group={{
                        id: 1,
                        name: 'Group 1',
                        link: '/group-1',
                        apps: [{ id: 1 }, { id: 3 }],
                    }}
                    addToGroup={jest.fn()}
                    i18n={mockI18n}
                />
            )
        ).toMatchSnapshot();
    });

    it('renders in all added mode', () => {
        expect(
            shallow(
                <AppGroup
                    appIds={[3, 4]}
                    group={{
                        id: 1,
                        name: 'Group 1',
                        link: '/group-1',
                        apps: [{ id: 3 }, { id: 4 }],
                    }}
                    addToGroup={jest.fn()}
                    i18n={mockI18n}
                />
            )
        ).toMatchSnapshot();
    });
});
