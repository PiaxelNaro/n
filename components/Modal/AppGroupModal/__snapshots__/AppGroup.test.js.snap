// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppGroup> renders enabled group 1`] = `
<AppGroup__StyledGroupWrapper>
  <FlexGrid>
    <FlexGrid.Item
      flex={1}
    >
      <FlexGrid>
        <Text__Text
          ellipsis={true}
        >
          <AppGroup__Link
            href="/group-1"
          >
            Group 1
          </AppGroup__Link>
        </Text__Text>
        <AppGroup__StyledGroupAppsCount>
          (2 apps)
        </AppGroup__StyledGroupAppsCount>
      </FlexGrid>
    </FlexGrid.Item>
    <div>
      <span
        className="action-context"
      >
        <MdControlPoint
          color="#3F88D4"
          size="16"
        />
      </span>
      <span
        className="action-button"
      >
        <ButtonSecondary
          onClick={[Function]}
        >
          Add To Group
        </ButtonSecondary>
      </span>
    </div>
  </FlexGrid>
</AppGroup__StyledGroupWrapper>
`;

exports[`<AppGroup> renders in all added mode 1`] = `
<AppGroup__StyledGroupWrapper>
  <FlexGrid>
    <FlexGrid.Item
      flex={1}
    >
      <FlexGrid>
        <Text__Text
          ellipsis={true}
        >
          <AppGroup__Link
            href="/group-1"
          >
            Group 1
          </AppGroup__Link>
        </Text__Text>
        <AppGroup__StyledGroupAppsCount>
          (2 apps)
        </AppGroup__StyledGroupAppsCount>
      </FlexGrid>
    </FlexGrid.Item>
    <div>
      <FlexGrid
        className="action-context"
        inline={true}
      >
        <AppGroup__StyledGroupAppsCount>
          All added
        </AppGroup__StyledGroupAppsCount>
        <FaCheckCircleO
          color="#00BFAA"
          size="16"
        />
      </FlexGrid>
      <span
        className="action-button"
      >
        <ButtonSecondary
          href="/group-1"
        >
          Go To Group
        </ButtonSecondary>
      </span>
    </div>
  </FlexGrid>
</AppGroup__StyledGroupWrapper>
`;

exports[`<AppGroup> renders in loading mode 1`] = `
<AppGroup__StyledGroupWrapper>
  <FlexGrid>
    <FlexGrid.Item
      flex={1}
    >
      <FlexGrid>
        <Text__Text
          ellipsis={true}
        >
          <AppGroup__Link
            href="/group-1"
          >
            Group 1
          </AppGroup__Link>
        </Text__Text>
        <AppGroup__StyledGroupAppsCount>
          (2 apps)
        </AppGroup__StyledGroupAppsCount>
      </FlexGrid>
    </FlexGrid.Item>
    <Spinner
      size={16}
    />
  </FlexGrid>
</AppGroup__StyledGroupWrapper>
`;

exports[`<AppGroup> renders in partially added mode 1`] = `
<AppGroup__StyledGroupWrapper>
  <FlexGrid>
    <FlexGrid.Item
      flex={1}
    >
      <FlexGrid>
        <Text__Text
          ellipsis={true}
        >
          <AppGroup__Link
            href="/group-1"
          >
            Group 1
          </AppGroup__Link>
        </Text__Text>
        <AppGroup__StyledGroupAppsCount>
          (2 apps)
        </AppGroup__StyledGroupAppsCount>
      </FlexGrid>
    </FlexGrid.Item>
    <div>
      <FlexGrid
        className="action-context"
        inline={true}
      >
        <AppGroup__StyledGroupAppsCount>
          1 already added
        </AppGroup__StyledGroupAppsCount>
        <MdControlPoint
          color="#3F88D4"
          size="16"
        />
      </FlexGrid>
      <span
        className="action-button"
      >
        <ButtonSecondary
          onClick={[Function]}
        >
          Add Remaining
        </ButtonSecondary>
      </span>
    </div>
  </FlexGrid>
</AppGroup__StyledGroupWrapper>
`;
